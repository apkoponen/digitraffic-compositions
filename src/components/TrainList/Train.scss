@import 'styles/_colors';

.Train {
  h2 {
    margin-top: 0;
  }
}

.Train__status {
  float: right;
  vertical-align: middle;
}

.Train__composition-wrapper {
  margin: -10px -10px 20px -10px;

  &:last-child {
    margin-bottom: 0;
  }
}