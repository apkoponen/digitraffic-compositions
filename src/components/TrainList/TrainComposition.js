import React, { Component } from 'react';
import Skeleton from '../Skeleton/Skeleton';
import './TrainComposition.css';

class TrainComposition extends Component {
  formatLocomotive = locomotives => {
    return locomotives
      .map(locomotive => `${locomotive.locomotiveType} ${locomotive.powerType}`)
      .join(', ');
  };

  formatWagons = (wagons, length) => {
    const formattedWagons = wagons
      .filter(wagon => wagon.wagonType)
      .map(wagon => wagon.wagonType)
      .join(', ');
    return `${formattedWagons} (${length} meters)`;
  };

  render() {
    const { composition, skeleton } = this.props;
    let locomotive, wagons, maxSpeed, originStation, destinationStation;
    if (skeleton) {
      locomotive = <Skeleton width="50" />;
      wagons = <Skeleton width="100" />;
      maxSpeed = <Skeleton width="50" />;
      originStation = <Skeleton width="25" />;
      destinationStation = <Skeleton width="25" />;
    } else {
      locomotive = this.formatLocomotive(composition.locomotives);
      wagons = this.formatWagons(composition.wagons, composition.totalLength);
      maxSpeed = composition.maximumSpeed + ' km/h';
      originStation = composition.beginTimeTableRow.stationShortCode;
      destinationStation = composition.endTimeTableRow.stationShortCode;
    }

    return (
      <div className="TrainComposition">
        <div className="TrainComposition__content">
          <p>
            <strong>Locomotive:</strong> {locomotive}
          </p>
          <p>
            <strong>Wagons:</strong> {wagons}
          </p>
          <p>
            <strong>Max speed:</strong> {maxSpeed}
          </p>
        </div>
        <div className="TrainComposition__timeline">
          <div className="TrainComposition__origin-station">
            {originStation}
          </div>
          <div className="TrainComposition__destination-station">
            {destinationStation}
          </div>
        </div>
      </div>
    );
  }
}

export default TrainComposition;
