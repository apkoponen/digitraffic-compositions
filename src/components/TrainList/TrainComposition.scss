@import 'styles/_colors';

$timeline-color: #666;

.TrainComposition {
  display: flex;
  align-items: stretch;

  p {
    margin-top: 0;
    margin-bottom: 5px;
  }
}

.TrainComposition__content {
  padding: 10px;
  font-size: 14px;
  flex: 1;
  background: #f6f6f6;
}

.TrainComposition__timeline {
  z-index: 2;
  flex: 0 0 30px;
  margin-left: 3px;
  border-left: 2px solid $timeline-color;
  position: relative;

  &:before, &:after {
    content: '';
    width: 8px;
    height: 8px;
    border-radius: 50%;
    position: absolute;
    left: -5px;
    background: $timeline-color;
    border: 2px solid #fff;
  }

  &:before {
    top: -2px;
  }

  &:after {
    bottom: -2px;
  }
}

.TrainComposition__origin-station, .TrainComposition__destination-station {
  position: absolute;
  left: 3px;
  color: #333;
  font-size: 12px;
}

.TrainComposition__origin-station {
  top: 0;
}

.TrainComposition__destination-station {
  bottom: 0;
}