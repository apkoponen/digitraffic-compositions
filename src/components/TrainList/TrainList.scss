@import 'styles/_colors';
@import 'styles/_shadows';

.TrainList {
  display: flex;
  align-items: stretch;
  flex-flow: row wrap;
  padding: 7.5px;
}

.TrainList__box {
  box-shadow: $box-shadow-normal;
  padding: 15px;
  margin: 7.5px;
  background: #fff;
  border-radius: 2px;
}

.TrainList__item {
  flex: 0 1 calc(33.3333% - 15px);
  @extend .TrainList__box;

  @media(max-width: 960px) {
    flex: 0 1 calc(50% - 15px);
  }

  @media(max-width: 640px) {
    flex: 0 1 100%;
  }
}

.Trainlist__empty-message {
  flex: 0 1 auto;
  text-align: center;
  @extend .TrainList__box;
}

.App__footer {
  text-align: center;
  padding: 15px;
}