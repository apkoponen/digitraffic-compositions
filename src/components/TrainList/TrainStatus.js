import React, { Component } from 'react';
import './TrainStatus.css';

class TrainStatus extends Component {
  render() {
    const statusClass = this.props.running ? 'running' : 'stopped';
    const statusText = this.props.running ? 'Moving' : 'Stopped';

    return (
      <div className={`TrainStatus ${statusClass}`}>
        <span className="TrainStatus__dot" />
        {statusText}
      </div>
    );
  }
}

export default TrainStatus;
