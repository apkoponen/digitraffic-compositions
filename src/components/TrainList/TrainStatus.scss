@import 'styles/_colors';

$color-running: #29a21b;
$color-stopped: #8B0000;

.TrainStatus {
  text-transform: uppercase;
  font-size: 12px;

  &.running {
    color: $color-running;
    .TrainStatus__dot {
      background: $color-running;
    }
  }
  &.stopped {
    color: $color-stopped;
    .TrainStatus__dot {
      background: $color-stopped;
    }
  }
}

.TrainStatus__dot {
  display: inline-block;
  width: 8px;
  height: 8px;
  vertical-align: baseline;
  border-radius: 50%;
  margin-right: 3px;
}