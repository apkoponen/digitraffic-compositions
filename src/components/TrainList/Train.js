import React, { Component } from 'react';
import TrainComposition from './TrainComposition';
import TrainStatus from './TrainStatus';
import './Train.css';

class Train extends Component {
  render() {
    const { train } = this.props;
    let compositions;
    if (train.compositionsLoaded) {
      if (train.compositions.length > 0) {
        compositions = train.compositions.map(composition =>
          <div
            className="Train__composition-wrapper"
            key={composition.beginTimeTableRow.scheduledTime}
          >
            <TrainComposition composition={composition} />
          </div>
        );
      } else {
        compositions = <div>No composition information available.</div>;
      }
    } else {
      compositions = <TrainComposition skeleton />;
    }

    return (
      <div className="Train">
        <div className="Train__status">
          <TrainStatus running={train.runningCurrently} />
        </div>
        <h2>
          {train.trainType}
          {train.trainNumber} {train.originStation}-{train.destinationStation}
        </h2>
        {compositions}
      </div>
    );
  }
}

export default Train;
