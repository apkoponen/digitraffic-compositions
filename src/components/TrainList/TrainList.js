import React, { Component } from 'react';
import Train from './Train';
import './TrainList.css';

class TrainList extends Component {
  render() {
    const trains = this.props.trains.map(train =>
      <div
        className="TrainList__item"
        key={train.trainNumber + train.departureDate}
      >
        <Train train={train} />
      </div>
    );
    return (
      <div className="TrainList">
        {trains}
      </div>
    );
  }
}

export default TrainList;
