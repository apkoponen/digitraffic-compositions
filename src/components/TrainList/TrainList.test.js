import React from 'react';
import TrainList from './TrainList';
import renderer from 'react-test-renderer';
import trains from '../../test-data/trains';

it('renders correctly', () => {
  const tree = renderer.create(<TrainList trains={trains} />).toJSON();
  expect(tree).toMatchSnapshot();
});
