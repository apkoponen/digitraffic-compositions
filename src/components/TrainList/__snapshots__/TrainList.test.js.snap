// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="TrainList"
>
  <div
    className="TrainList__item"
  >
    <div
      className="Train"
    >
      <div
        className="Train__status"
      >
        <div
          className="TrainStatus stopped"
        >
          <span
            className="TrainStatus__dot"
          />
          Stopped
        </div>
      </div>
      <h2>
        IC
        943
         
        HKI
        -
        TKU
      </h2>
      <div
        className="Train__composition-wrapper"
      >
        <div
          className="TrainComposition"
        >
          <div
            className="TrainComposition__content"
          >
            <p>
              <strong>
                Locomotive:
              </strong>
               
              Sr2 Electric
            </p>
            <p>
              <strong>
                Wagons:
              </strong>
               
              Edfs, Edb, Edo (99 meters)
            </p>
            <p>
              <strong>
                Max speed:
              </strong>
               
              160 km/h
            </p>
          </div>
          <div
            className="TrainComposition__timeline"
          >
            <div
              className="TrainComposition__origin-station"
            >
              HKI
            </div>
            <div
              className="TrainComposition__destination-station"
            >
              TKU
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="TrainList__item"
  >
    <div
      className="Train"
    >
      <div
        className="Train__status"
      >
        <div
          className="TrainStatus stopped"
        >
          <span
            className="TrainStatus__dot"
          />
          Stopped
        </div>
      </div>
      <h2>
        IC
        984
         
        TKU
        -
        HKI
      </h2>
      <div
        className="Train__composition-wrapper"
      >
        <div
          className="TrainComposition"
        >
          <div
            className="TrainComposition__content"
          >
            <p>
              <strong>
                Locomotive:
              </strong>
               
              Sr2 Electric
            </p>
            <p>
              <strong>
                Wagons:
              </strong>
               
              Edo, Edb, Edfs (99 meters)
            </p>
            <p>
              <strong>
                Max speed:
              </strong>
               
              160 km/h
            </p>
          </div>
          <div
            className="TrainComposition__timeline"
          >
            <div
              className="TrainComposition__origin-station"
            >
              TKU
            </div>
            <div
              className="TrainComposition__destination-station"
            >
              HKI
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="TrainList__item"
  >
    <div
      className="Train"
    >
      <div
        className="Train__status"
      >
        <div
          className="TrainStatus stopped"
        >
          <span
            className="TrainStatus__dot"
          />
          Stopped
        </div>
      </div>
      <h2>
        IC
        943
         
        HKI
        -
        TKU
      </h2>
      <div
        className="Train__composition-wrapper"
      >
        <div
          className="TrainComposition"
        >
          <div
            className="TrainComposition__content"
          >
            <p>
              <strong>
                Locomotive:
              </strong>
               
              Sr2 Electric
            </p>
            <p>
              <strong>
                Wagons:
              </strong>
               
              Edfs, Edb, Edo (99 meters)
            </p>
            <p>
              <strong>
                Max speed:
              </strong>
               
              160 km/h
            </p>
          </div>
          <div
            className="TrainComposition__timeline"
          >
            <div
              className="TrainComposition__origin-station"
            >
              HKI
            </div>
            <div
              className="TrainComposition__destination-station"
            >
              TKU
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="TrainList__item"
  >
    <div
      className="Train"
    >
      <div
        className="Train__status"
      >
        <div
          className="TrainStatus stopped"
        >
          <span
            className="TrainStatus__dot"
          />
          Stopped
        </div>
      </div>
      <h2>
        IC
        984
         
        TKU
        -
        HKI
      </h2>
      <div>
        No composition information available.
      </div>
    </div>
  </div>
</div>
`;
